File: src/main/java/december_2024/NonOverlappingIntervals.java
package december_2024;

import java.util.Arrays;
import java.util.Comparator;

public class NonOverlappingIntervals {

    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));

        int res = 0;
        int prevEnd = intervals[0][1];
        for (int i = 1; i < intervals.length; i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            if (start >= prevEnd) {
                prevEnd = end;
            } else {
                res++;
                prevEnd = Math.min(end, prevEnd);
            }
        }
        return res;
    }

}

File: python/non_overlapping_intervals.py
class Solution(object):
    # O(n log()) time | O(n) space
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        intervals.sort()
        prevEnd = intervals[0][1]
        res = 0
        for i in range(1, len(intervals)):
            start = intervals[i][0]
            end = intervals[i][1]
            if start >= prevEnd:
                prevEnd = end
            else:
                res += 1
                prevEnd = min(prevEnd, end)
        return res


intervals = [[1,2],[2,3],[3,4],[1,3]]

solution = Solution()
print(solution.eraseOverlapIntervals(intervals))